### ISSUE TITLE
Various issues with bitstring types

### ISSUE THREAD

**Issue by trarbr** [Qqwy/elixir-type_check#104]
Hi

First of all, thank you for all your hard work on this excellent library :)

I am experimenting with the use of TypeCheck for specifying types and specs for a library that does a lot of binary decoding and encoding (Blue Heron, a Bluetooth library). I would like to bring in TypeCheck for documentation, runtime checks and test data generation.

This is my first time really using TypeCheck, and I have run into a few issues. I am not sure if it's because I am doing it wrong, or because of bugs. Take the following code as an example:

```elixir
defmodule Diacheck.Frame do
  use TypeCheck

  @type! status() :: byte()

  @spec! decode(<<_::8>>) :: status()
  def decode(<<status>>) do
    status
  end
end
```

The decode function extracts an integer in range 0..255 from the input, which must be a bitstring with a length 8 bits.

If I run `Diacheck.Frame.decode(<<40>>)`, that works fine. But if I run `Diacheck.Frame.decode(<<40, 1>>)` I get the following error:

```
** (Protocol.UndefinedError) protocol String.Chars not implemented for {:doc_cons, {:doc_color, "8", :yellow}, {:doc_color, :doc_nil, :red}} of type Tuple. This protocol is implemented for the following type(s): Atom, BitString, Date, DateTime, Float, Integer, List, NaiveDateTime, Time, URI, Version, Version.Requirement
    (elixir 1.13.2) lib/string/chars.ex:3: String.Chars.impl_for!/1
    (elixir 1.13.2) lib/string/chars.ex:22: String.Chars.to_string/1
    (type_check 0.10.7) lib/type_check/builtin/sized_bitstring.ex:45: TypeCheck.Protocols.Inspect.TypeCheck.Builtin.SizedBitstring.inspect/2
    (type_check 0.10.7) lib/type_check/protocols/inspect.ex:101: TypeCheck.Inspect.inspect/2
    (type_check 0.10.7) lib/type_check/protocols/inspect.ex:118: TypeCheck.Inspect.inspect_binary/2
    (type_check 0.10.7) lib/type_check/type_error/default_formatter.ex:267: TypeCheck.TypeError.DefaultFormatter.do_format/1
    (type_check 0.10.7) lib/type_check/type_error/default_formatter.ex:6: TypeCheck.TypeError.DefaultFormatter.format/2
    (type_check 0.10.7) lib/type_check/type_error.ex:57: TypeCheck.TypeError.exception/1
```

Also, if I run `mix dialyzer` in the project, it complains like this:

```
Compiling 4 files (.ex)
Generated diacheck app
Finding suitable PLTs
Checking PLT...
[:compiler, :elixir, :iex, :kernel, :logger, :stdlib, :type_check]
Looking up modules in dialyxir_erlang-24.2_elixir-1.13.2_deps-dev.plt
Looking up modules in dialyxir_erlang-24.2_elixir-1.13.2.plt
Finding applications for dialyxir_erlang-24.2_elixir-1.13.2.plt
Finding modules for dialyxir_erlang-24.2_elixir-1.13.2.plt
Checking 449 modules in dialyxir_erlang-24.2_elixir-1.13.2.plt
Finding applications for dialyxir_erlang-24.2_elixir-1.13.2_deps-dev.plt
Finding modules for dialyxir_erlang-24.2_elixir-1.13.2_deps-dev.plt
Copying dialyxir_erlang-24.2_elixir-1.13.2.plt to dialyxir_erlang-24.2_elixir-1.13.2_deps-dev.plt
Looking up modules in dialyxir_erlang-24.2_elixir-1.13.2_deps-dev.plt
Checking 449 modules in dialyxir_erlang-24.2_elixir-1.13.2_deps-dev.plt
Adding 335 modules to dialyxir_erlang-24.2_elixir-1.13.2_deps-dev.plt
done in 0m54.34s
No :ignore_warnings opt specified in mix.exs and default does not exist.

Starting Dialyzer
[
  check_plt: false,
  init_plt: '/Users/trarbr/diacheck/_build/dev/dialyxir_erlang-24.2_elixir-1.13.2_deps-dev.plt',
  files: ['/Users/trarbr/diacheck/_build/dev/lib/diacheck/ebin/Elixir.Diacheck.Frame.beam',
   '/Users/trarbr/diacheck/_build/dev/lib/diacheck/ebin/Elixir.Diacheck.FrameOld.beam',
   '/Users/trarbr/diacheck/_build/dev/lib/diacheck/ebin/Elixir.Diacheck.Type.beam',
   '/Users/trarbr/diacheck/_build/dev/lib/diacheck/ebin/Elixir.Diacheck.beam',
   '/Users/trarbr/diacheck/_build/dev/lib/diacheck/ebin/Elixir.TypeCheck.Internals.UserTypes.Diacheck.Frame.beam',
   ...],
  warnings: [:unknown]
]
Total errors: 1, Skipped: 0, Unnecessary Skips: 0
done in 0m0.39s
lib/type_check/spec.ex:1:call_without_opaque
Function call without opaqueness type mismatch.

Call does not have expected term of type
  {{TypeCheck.Builtin.Map.t()
    | %{
        :__struct__ => atom(),
        :choices => [any()],
        :element_type => _,
        :element_types => [any()],
        :keypairs => [any()],
        :local => boolean(),
        :name => atom(),
        :range => map(),
        :type => _,
        :value => _
      }, atom(),
    %{
      :expected_length => non_neg_integer(),
      :expected_size => integer(),
      :index => integer(),
      :key => _,
      :keys => [any()],
      :problem => _,
      :problems => [any()]
    }, _}, nil | maybe_improper_list() | map()}
  | {TypeCheck.Builtin.Map.t()
     | %{
         :__struct__ => atom(),
         :choices => [any()],
         :element_type => _,
         :element_types => [any()],
         :keypairs => [any()],
         :local => boolean(),
         :name => atom(),
         :range => map(),
         :type => _,
         :value => _
       }, atom(),
     %{
       :expected_length => non_neg_integer(),
       :expected_size => integer(),
       :index => integer(),
       :key => _,
       :keys => [any()],
       :problem => _,
       :problems => [any()]
     }, _}
 (with opaque subterms) in the 1st position.

TypeCheck.TypeError.exception(
  {{%TypeCheck.Spec{
      :location => {<<_::528>>, 6},
      :name => :decode,
      :param_types => [
        %TypeCheck.Builtin.SizedBitstring{:prefix_size => 8, :unit_size => nil},
        ...
      ],
      :return_type => %TypeCheck.Builtin.NamedType{
        :local => false,
        :name => <<_::184>>,
        :type => %TypeCheck.Builtin.Range{:range => map()}
      }
    }, :param_error,
    %{
      :index => 0,
      :problem =>
        {%TypeCheck.Builtin.SizedBitstring{:prefix_size => 8, :unit_size => nil},
         :no_match | :wrong_size, %{}, _}
    }, [any(), ...]}, [{:file, <<_::528>>} | {:line, 1}, ...]}
)

________________________________________________________________________________
done (warnings were emitted)
Halting VM with exit status 2
```

If I remove the bang from spec!, `mix dialyzer` finds no errors.

I also noted one more thing. If I specify the binary literal in the wrong way (e.g. `<<_::_>>` instead of `<<_::8>>`), the error says that the `size` parameter specifies the byte-size of he bitstring, but I do believe it should be the bit-size of the bitstring. 

```
== Compilation error in file lib/diacheck/frame.ex ==
** (TypeCheck.CompileError) TypeCheck does not support the bitstring literal `<<_::_>>`
Currently supported are:
- <<>> -> empty bitstring
- <<_ :: size >> -> a bitstring of exactly `size` bytes long <---- This should say bits long instead bytes long
- <<_ :: _ * unit >> -> a bitstring whose length is divisible by `unit`.
- <<_ :: size, _ * unit >> -> a bitstring whose (length - `size`) is divisible by `unit`.

    (type_check 0.10.7) lib/type_check/internals/pre_expander.ex:137: TypeCheck.Internals.PreExpander.rewrite/3
    (type_check 0.10.7) lib/type_check/type.ex:81: TypeCheck.Type.build_unescaped/4
    (elixir 1.13.2) lib/enum.ex:1593: Enum."-map/2-lists^map/1-0-"/2
    (type_check 0.10.7) lib/type_check/macros.ex:236: anonymous fn/3 in TypeCheck.Macros.create_spec_defs/3
    (elixir 1.13.2) lib/enum.ex:2396: Enum."-reduce/3-lists^foldl/2-0-"/3
    (type_check 0.10.7) lib/type_check/macros.ex:232: TypeCheck.Macros.create_spec_defs/3
    (type_check 0.10.7) expanding macro: TypeCheck.Macros.__before_compile__/1
    lib/diacheck/frame.ex:1: Diacheck.Frame (module)
```

**Comment by Qqwy** [comment]
Thank you very much for this bug report, as well as fixing the two smaller issues in the PR you contributed! :heart_eyes: 

I will dig deeper into why Dialyzer is not happy with the generated spec code. The error message looks very daunting, but there probably is a simple root cause.

**Comment by Qqwy** [comment]
Very interesting. I am able to reproduce the dialyzer error on Elixir 1.13.2 / OTP 24. However, on Elixir 1.12.0 / OTP 24 dialyzer does not raise any errors for the same code.

:thinking:

**Comment by Qqwy** [comment]
I'm pretty sure that this is a particular case of the Dialyzer error(s) in #95.

**Comment by trarbr** [comment]
Thank you for looking into it so quickly 👍  and releasing the inspect fix! I hope you can find a way to solve #95, but it's not a blocker for me, so no worries. Unfortunately I have no idea how to help with it. Feel free to close this issue and merge it into the other one if you want.

**Comment by Qqwy** [comment]
Browsing through the issues again; This one should definitely be closed in favour of #95 .
